cli/cmd/edges.go

```diff
@@ -0,0 +1,297 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/linkerd/linkerd2/controller/api/util"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type indexedEdgeResults struct {
+	ix   int
+	rows []*pb.Edge
+	err  error
+}
+
+func newCmdEdges() *cobra.Command {
+	options := newStatOptions()
+
+	cmd := &cobra.Command{
+		Use:   "edges [flags] (RESOURCETYPE)",
+		Short: "Display connections between resources, and the identity of their associated Linkerd proxies (if known)",
+		Long: `Display connections between resources, and the identity of their associated Linkerd proxies (if known).
+
+  The RESOURCETYPE argument specifies the type of resource to display edges within. A namespace must be specified.
+
+  Examples:
+  * deploy
+  * ds
+  * job
+	* po
+	* rc
+	* sts
+
+  Valid resource types include:
+  * daemonsets
+  * deployments
+  * jobs
+  * pods
+  * replicationcontrollers
+  * statefulsets
+
+If no resource name is specified, displays edges within all resources of the specified RESOURCETYPE`,
+		Example: `  # Get all edges between pods in the test namespace.
+  linkerd edges deploy -n test`,
+		Args:      cobra.MinimumNArgs(1),
+		ValidArgs: util.ValidTargets,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			reqs, err := buildEdgesRequests(args, options)
+			if err != nil {
+				return fmt.Errorf("error creating metrics request while making edges request: %v", err)
+			}
+
+			// The gRPC client is concurrency-safe, so we can reuse it in all the following goroutines
+			// https://github.com/grpc/grpc-go/issues/682
+			client := checkPublicAPIClientOrExit()
+			c := make(chan indexedEdgeResults, len(reqs))
+			for num, req := range reqs {
+				go func(num int, req *pb.EdgesRequest) {
+					resp, err := requestEdgesFromAPI(client, req)
+					if err != nil {
+						fmt.Println(err)
+					}
+
+					rows := edgesRespToRows(resp)
+
+					c <- indexedEdgeResults{num, rows, err}
+				}(num, req)
+			}
+
+			totalRows := make([]*pb.Edge, 0)
+			i := 0
+			for res := range c {
+				if res.err != nil {
+					return res.err
+				}
+				totalRows = append(totalRows, res.rows...)
+				if i++; i == len(reqs) {
+					close(c)
+				}
+			}
+
+			output := renderEdgeStats(totalRows, options)
+			_, err = fmt.Print(output)
+
+			return err
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified resource")
+	cmd.PersistentFlags().StringVarP(&options.outputFormat, "output", "o", options.outputFormat, "Output format; one of: \"table\" or \"json\"")
+	return cmd
+}
+
+func buildEdgesRequests(resources []string, options *statOptions) ([]*pb.EdgesRequest, error) {
+	targets, err := util.BuildResources(options.namespace, resources)
```

Comment:
unless we're planning to support things like `deploy/my-deploy`, we probably want to validate that `target.Name` is empty, somewhere around here.